IdentityCore/src/cache/mac/MSIDMacCredentialStorageItem.m

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+
+#import "MSIDMacCredentialStorageItem.h"
+
+static NSString *keyDelimiter = @"-";
+
+@interface MSIDMacCredentialStorageItem ()
+
+@property (nonatomic) NSMutableDictionary *cacheObjects;
+@property (nonatomic) dispatch_queue_t queue;
+
+@end
+
+@implementation MSIDMacCredentialStorageItem
+
+- (instancetype)init
+{
+    if(self = [super init])
+    {
+        self.cacheObjects = [NSMutableDictionary dictionary];
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.universalstorage-%@", [NSUUID UUID].UUIDString];
+        self.queue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+    }
+    
+    return self;
+}
+
+- (void)storeCredential:(MSIDCredentialCacheItem *)credential forKey:(NSString *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects setObject:credential forKey:key];
+    });
+}
+
+- (void)mergeStorageItem:(MSIDMacCredentialStorageItem *)storageItem
+{
+    NSArray *keys = [storageItem allKeys];
+    for (NSString *key in keys)
+    {
+        if (![self storedCredentialForKey:key])
+        {
+            MSIDCredentialCacheItem *credential = [storageItem.cacheObjects objectForKey:key];
```

Comment:
should this be copied? what if other storage item gets modified at the same time (e.g. this credential gets removed from another storage item)

Reply:
This should be in stack and I don't think it needs to be copied as its a value in the dictionary.

Reply:
I looked at this code again, and I still think someone might modify the dictionary or the item while we're processing it here. 

I'm not fully understanding what you meant by your comment. Obj-C objects are always allocated on heap (except blocks in certain scenarios). See [this](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/WorkingwithObjects/WorkingwithObjects.html).

Multiple threads can be modifying same underlying dictionary while it's getting processed here. Somebody can also modify the actual value from the dictionary(credential cache item) that new dictionary will point to at a later point. Therefore, I'm still thinking that this code is not safe.

Consider following sample code (it showcases modification without multiple threads):

```
- (void)callMe
{
    NSMutableDictionary *dict = [NSMutableDictionary new];
    dict[@"test1"] = @"test2";
    dict[@"test3"] = @"test4";
    [self myMethod:dict modifyBlock:^{
        dict[@"newtest"] = @"test5";
    }];
}

- (void)myMethod:(NSDictionary *)dictionary modifyBlock:(void (^)(void))callback
{
    NSLog(@"Dictionary: %@", dictionary);
    callback();
    NSLog(@"Dictionary: %@", dictionary);
}
```

Reply:
There are 2 variables in the mergeStorageItem code. There is local object reference which we retrieve from keychain (savedStorageItem) which which can only be accessed by the stack frame of the method and is thread safe. Another variable is the currentStorage item which is shared by threads and can be modified while the mergeStorageItem code is in progress and cause issues. 

I think copying the values of the local object here will not prevent this issue rather we should prevent the shared item being modified while the merge is in progress.

I think there are 2 options to do this, either copy the entire shared dictionary before the merge or synchronize the entire merge operation using dispatch calls. The same 2 options should also be applied to the serialize method as it can have the same issue. I will update the PR with the change and we can review it then.

For the issue you mentioned regarding copying a object passed by reference to prevent it from being modified, I would rather prefer the caller to do that by creating a mutable copy. If we copy all the values in the method itself, it would be causing similar performance issues we discussed using deep copy.:)

We can discuss this further in person.